File: src/main/java/shopping/application/product/ProductService.java
package shopping.application.product;

import org.springframework.transaction.annotation.Transactional;
import org.springframework.stereotype.Service;
import shopping.domain.product.Price;
import shopping.domain.product.Product;
import shopping.domain.product.ProductName;
import shopping.domain.product.ProfanityChecker;
import shopping.domain.product.exception.ProductNotFoundException;
import shopping.domain.product.exception.ProfanityNameException;
import shopping.domain.repository.ProductRepository;
import shopping.dto.ProductRequest;
import shopping.dto.ProductResponse;

import java.util.List;

@Service
public class ProductService {
    private final ProfanityChecker profanityChecker;
    private final ProductRepository productRepository;

    public ProductService(ProfanityChecker profanityChecker, ProductRepository productRepository) {
        this.profanityChecker = profanityChecker;
        this.productRepository = productRepository;
    }

    @Transactional
    public Long createProduct(ProductRequest request) {
        if(profanityChecker.containsProfanity(request.getName())) {
            throw new ProfanityNameException(request.getName());
        }

        Product product = Product.create(request.getName(), request.getPrice(), request.getImageUrl());

        Product savedProduct = productRepository.save(product);
        return savedProduct.getId();
    }

    @Transactional
    public void updateProduct(Long id, ProductRequest request) {
        Product product = productRepository.findById(id)
                .orElseThrow(ProductNotFoundException::new);

        if(profanityChecker.containsProfanity(request.getName())) {
            throw new ProfanityNameException(request.getName());
        }

        ProductName newName = new ProductName(request.getName());
        Price newPrice = new Price(request.getPrice());

        product.update(newName, newPrice, request.getImageUrl());
    }

    @Transactional(readOnly = true)
    public ProductResponse getProduct(Long id) {
        Product product = productRepository.findById(id)
                .orElseThrow(ProductNotFoundException::new);

        return ProductResponse.from(product);
    }

    @Transactional(readOnly = true)
    public List<ProductResponse> getAllProducts() {
        return productRepository.findAll().stream()
                .map(ProductResponse::from)
                .toList();
    }
}

File: src/main/kotlin/shopping/application/product/ProductService.kt
package shopping.application.product

import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import shopping.domain.product.Product
import shopping.domain.product.ProductBadWordClient
import shopping.domain.product.ProductRepository
import shopping.support.error.CoreException
import shopping.support.error.ErrorType
import org.springframework.stereotype.Service
import java.math.BigDecimal

@Service
class ProductService(
    private val productRepository: ProductRepository,
    private val productBadWordClient: ProductBadWordClient
) {
    fun findByPaging(page: Int, size: Int): Page<Product> {
        return productRepository.findAll(PageRequest.of(page, size))
    }

    fun findById(id: Long): Product {
        return productRepository.findById(id)
            ?: throw CoreException(ErrorType.NOT_FOUND, "상품이 존재하지 않습니다.")
    }

    fun create(name: String, price: BigDecimal, imageUrl: String): Product {
        validateProfanity(name)
        val product = Product(
            name = name,
            price = price,
            imageUrl = imageUrl
        )
        return productRepository.save(product)
    }

    fun update(id: Long, name: String, price: BigDecimal, imageUrl: String): Product {
        validateProfanity(name)
        val product = findById(id)
        product.update(name, price, imageUrl)
        return productRepository.save(product)
    }

    private fun validateProfanity(name: String) {
        if (productBadWordClient.containsProfanity(name)) {
            throw CoreException(ErrorType.BAD_REQUEST, "상품 이름에 비속어를 포함할 수 없습니다.")
        }
    }

    fun delete(id: Long) {
        val product = findById(id)
        productRepository.deleteById(product.id)
    }
}
